### PR TITLE
Update phpstan to the most recent version

### PR DESCRIPTION
**Pull request by usox**
This updates phpstan and introduces a bunch of new type hints and type corrections. Also add phpstan-strict-rules to be even more strict.

- Bump minimal required php version to 7.2
- Update/correct/add some type hints in the tests

### CODE REVIEW

**Comment by usox** [comment]
Ah, it's 7.1 again - sigh. Well I guess, downgrading stan to 1.4 wouldn't be much of a problem.

**Comment by tuupola** [comment]
Oh wait. Actually I think we can do an exception and allow PHP 7.1 to fail, or alternatively bump up the minimum requirements 7.1 has been EOL for a while...

**Comment by usox** [comment]
> Oh wait. Actually I think we can do an exception and allow PHP 7.1 to fail, or alternatively bump up the minimum requirements 7.1 has been EOL for a while...

I already prepared a follow up PR which bumps neomerx/cors-psr7 to v2. I'd be happy to also upgrade the minimum php version requirement to...well...what? 7.4 (which is also EOL)...so maybe 8?

**Comment by tuupola** [comment]
Only PHP 7.1 failed so only it should be removed.

**Comment by usox** [comment]
Here we go.

File: src/CorsMiddleware.php [hunk 1]
```diff
@@ -167,19 +211,19 @@ private function buildSettings(ServerRequestInterface $request, ResponseInterfac
...
             $methods = (array) $this->options["methods"]($request, $response);
         } else {
-            $methods = $this->options["methods"];
+            $methods = (array) $this->options["methods"];
         }
         $methods = array_fill_keys($methods, true);
         $settings->setRequestAllowedMethods($methods);
 
         $headers = array_fill_keys($this->options["headers.allow"], true);
-        $headers = array_change_key_case($headers, CASE_LOWER);
```
Comment (tuupola) [inline src/CorsMiddleware.php:182]: Did PHPStan complain about this? I prefer having the CASE_LOWER there so it is easy to know what is happening without going to PHP docs to see what is the default behaviour.

Comment (usox) [reply src/CorsMiddleware.php:182]: It did not (but rector will).
I added a comment line above to describe what's happening in here. Adding default-parameters just to document the methods behavior seems a bit odd to me. In IDEs, the methods signature is just a single click away.

**Comment by tuupola** [comment]
Thanks!

**Pull request merged**